### PR TITLE
Assure compatibility with FIX Orchestra generated QFJ dictionary

### LINKED ISSUE #336: A test is needed to defend QuickFIXJ compatibility with FIX Dictionary Files generated from FIX Orchestra
The FIX Orchestra project includes a utility to generate QFJ compatible dictionaries from an Orchestra Repo https://github.com/FIXTradingCommunity/fix-orchestra

The Orchestra model and the QuickFIX model are inconsistent in the way that they deal with Components and Groups.
QuickFIX does not normalise Groups, Groups are defined in situ within the containing component.
There are a small number of cases where Groups are required. 
To represent this without requiring that the Component and Group be defined in situ in the message the Group must be defined as required within the Component. e.g. ```<group name="NoPartyIDs" required="Y"> ```
There are cases such as Component "Parties" where the component is not always required.
QuickFIXJ validation does behave correctly in this case:
* given that the Component is not required when the contained Group is required and the Component is absent then validation succeeds 
* given that the Component is required when the contained Group is required and the Component is absent then validation fails
* given that the Component is required when the contained Group is required and the Component is present then validation succeeds

Providing a test case will guard against possible regressions that would introduce incompatibility between FIX Orchestra tooling and QuickFIXJ.

Please see also 👍 
https://github.com/FIXTradingCommunity/fix-orchestra/issues/104

### PR DESCRIPTION
**Pull request by david-gibbs-ig**
Fixes #336 

Tests to assure compatibility with FIX Orchestra generated QFJ dictionary : issue-336

see https://github.com/quickfix-j/quickfixj/issues/336

### CODE REVIEW

**Comment by david-gibbs-ig** [comment]
@chrjohn Is there a  way to rerun failed checks , I'm pretty sure that concurrency problem in the JDK15 check is not due to this change :-)

**Comment by chrjohn** [comment]
Can't you re-run the checks from the Checks tab? But maybe you are not allowed to do so. 
Otherwise close and reopen the PR should work but is of course ugly.

**Comment by david-gibbs-ig** [comment]
``` git commit -m "retrigger checks" --allow-empty && git push <remote> HEAD```  I will try the close and re-open next time.

**Comment by david-gibbs-ig** [comment]
@chrjohn  I think that there are instabilities in the test cases. perhaps on windows specifically ? Do you see this concurrent modification exception in other builds ?

**Comment by chrjohn** [comment]
The concurrent modification problems should get corrected by #324

File: quickfixj-core/src/test/java/quickfix/DataDictionaryTest.java [hunk 1]
```diff
@@ -1027,7 +1029,80 @@ public void testOrderedFields() throws Exception {
         int[] expectedPartySubIDsFieldOrder = new int[] {PartySubID.FIELD, PartySubIDType.FIELD};
         assertArrayEquals(expectedPartySubIDsFieldOrder, partySubIDsDictionary.getOrderedFields());
     }
+    
+    private QuoteRequest.NoRelatedSym.NoPartyIDs createPartiesGroup() {
+		final  QuoteRequest.NoRelatedSym.NoPartyIDs noPartiesGroup = new QuoteRequest.NoRelatedSym.NoPartyIDs();
+	    noPartiesGroup.set(new PartyID("PartyOnDude"));
```
Comment (chrjohn) [inline quickfixj-core/src/test/java/quickfix/DataDictionaryTest.java:1035]: :+1:

**Comment by chrjohn** [comment]
@david-gibbs-ig , could you please check the formatting on this change? It seems the indentation is sometimes two spaces, sometimes four...
Thanks in advance.

**Comment by david-gibbs-ig** [comment]
I've have converted the tabs to spaces, also updated Apache Mina version to address a security patch.  See the release notes here https://mina.apache.org/mina-project/.

**Comment by david-gibbs-ig** [comment]
This will be obsolete.

**Pull request closed without merging**